controllers/etcdcluster_controller.go

```diff
@@ -86,51 +115,131 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
+				"desired-peers", cluster.Spec.Replicas,
 				"peer", peerName)
-			return ctrl.Result{}, err
+			peer := peerForCluster(cluster, peerName)
+			if err := r.Create(ctx, peer); err != nil {
+				log.Error(err, "Failed to create peer",
+					"peer", peerName)
+				return ctrl.Result{}, nil, err
+			}
+			return ctrl.Result{}, &event.PeerCreatedEvent{Object: cluster, PeerName: peer.Name}, nil
 		}
-		return ctrl.Result{}, nil
+	} else {
+		log.V(2).Info("Cluster communication established")
 	}
+	return ctrl.Result{RequeueAfter: time.Second * 10}, nil, nil
+}
 
-	return ctrl.Result{}, nil
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=core,resources=endpoints,verbs=get;list
+
+func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	log := r.Log.WithValues("cluster", req.NamespacedName)
+
+	cluster := &etcdv1alpha1.EtcdCluster{}
+	if err := r.Get(ctx, req.NamespacedName, cluster); err != nil {
+		log.Error(err, "unable to fetch EtcdCluster resource")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+		// special 'exit early' case.
+	}
+
+	// Attempt to dial the etcd cluster, recording the cluster response if we can
+	members, err := r.connectToEtcd(ctx, cluster)
+	if err != nil {
+		log.Error(err, "Unable to contact etcd cluster")
+	}
+
+	// Perform a reconcile, getting back the desired result, any errors, and a clusterEvent. This is an internal concept
+	// and is not the same as the Kubernetes event, although it is used to make one later.
+	result, clusterEvent, err := r.reconcile(ctx, req.NamespacedName, members, cluster)
+	if err != nil {
+		log.Error(err, "Failed to reconcile")
+	}
+
+	// The update status takes in the cluster definition, and the member list from etcd as of *before we ran reconcile*.
+	// We also get the event, which may contain rich information about what we did (such as the new member name on a
+	// MemberAdded event).
+	err = r.updateStatus(ctx, req.NamespacedName, cluster, members, clusterEvent)
+	if err != nil {
+		log.Error(err, "Failed to update status")
+	}
+
+	// Finally, the event is used to generate a Kubernetes event by calling `Record` and passing in the recorder.
+	if clusterEvent != nil {
+		clusterEvent.Record(r.Recorder)
+	}
+
+	return result, nil
+}
+
+func (r *EtcdClusterReconciler) connectToEtcd(ctx context.Context, cluster *etcdv1alpha1.EtcdCluster) ([]etcdclient.Member, error) {
+	etcdConfig := etcdClientConfig(cluster)
+	c, err := etcdclient.New(*etcdConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to connect to etcd")
```

Comment:
I stopped using `errors.Wrap` in favour of the new `fmt.Errorf("%w", err)` feature:  https://blog.golang.org/go1.13-errors
But I don't think it gives you tracebacks, like the errors package.